### DISCUSSION TITLE
Generate a truncated cone with IfcOpenShell-Python

### DISCUSSION THREAD

**Discussion by michael-ziegler-ost** [IfcOpenShell/IfcOpenShell#4602] [General]
I'm pretty new to the python-library IfcOpenShell working on a project. I'm currently trying to generate a "truncated cone" from code (Python) to safe it afterwards as an IFC-file. One attempt that did work (but obviously not a desired solution) is to approximate the truncated cone with a lot of cylinders that are merged to a representation.

Is there a better solution to my problem? I found the element "IfcSurfaceOfRevolution", but I'm not able to generate the truncated cone with this element (the idea would be to use a trapezoidal-shaped profile in combination with the IfcSurfaceOfRevolution. Any experience or idea?

Many thanks in advance.

**Comment by theoryshaw** [comment]
Could you use this?
https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcSectionedSolidHorizontal.htm

**Reply by michael-ziegler-ost** [reply]
Thank you very much for your fast reaction!

"The solid is generated by sweeping the CrossSections between CrossSectionPositions with linear interpolation between profile points with the same tag along the directrix." sounds promising. But as it is mentioned, the element is not part of the standard - which is a precondition of my project.

It's not possible with IfcSurfaceOfRevolution to create the truncated cone?

**Reply by theoryshaw** [reply]
> It's not possible with IfcSurfaceOfRevolution to create the truncated cone?

yes, i think that could work.  What schema are you using? ifc2x3?

**Reply by michael-ziegler-ost** [reply]
I'm using schema IFC4... the code I'm using at the moment:

`builder = ifcopenshell.util.shape_builder.ShapeBuilder(model)

body = run("context.add_context", model, context_type="Model")

outer_curve = builder.polyline([(0.,0.),(2.,0.),(1.,3.),(0.,3.)], closed=True)
profile = builder.profile(outer_curve, "truncated_cone")
axis = model.create_entity("IfcAxis1Placement")
truncated_cone_profile = model.create_entity("IfcSurfaceOfRevolution", SweptCurve=profile, AxisPosition=axis)
representation = model.create_entity("IfcShapeRepresentation", RepresentationIdentifier="Body", RepresentationType="CSG", Items=[truncated_cone_profile])
product = run("root.create_entity", model, ifc_class="IFCBUILDINGELEMENTPROXY", name="Test")
run("geometry.assign_representation", model, product=product, representation=representation)
run("geometry.edit_object_placement", model, product=product)
run("spatial.assign_container", model, relating_structure=storey, product=product)`

unfortunately, something is wrong - it creates the element in the IFC file, but opening the file with a viewer ends in an error message....

**Reply by theoryshaw** [reply]
sorry not sure.

**Reply by michael-ziegler-ost** [reply]
many thanks nevertheless

**Comment by theoryshaw** [comment]
upon closer inspection it looks like these 2 methods don't have python wrappers yet.  Unless I'm missing something.

**Comment by Moult** [comment]
You can use IfcRightCircularCone then IfcBooleanResult with IfcHalfSpaceSolid.

**Reply by michael-ziegler-ost** [reply]
Thank you a lot for your hint - I'm going to try it that way.
